core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/metrics/BytesReadTrackingPublisher.java

```diff
@@ -16,52 +16,75 @@
 package software.amazon.awssdk.core.internal.metrics;
 
 import java.nio.ByteBuffer;
+import java.util.Optional;
 import java.util.concurrent.atomic.AtomicLong;
 import org.reactivestreams.Publisher;
 import org.reactivestreams.Subscriber;
 import org.reactivestreams.Subscription;
 import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.internal.util.ProgressUpdaterInvoker;
+import software.amazon.awssdk.http.async.SdkHttpContentPublisher;
 
 /**
  * Publisher that tracks how many bytes are published from the wrapped publisher to the downstream subscriber.
+ * If request contains Progress Listeners attached, the callbacks invoke methods to update and track request status
+ * by invoking progress updater methods with the bytes being transacted
  */
 @SdkInternalApi
-public final class BytesReadTrackingPublisher implements Publisher<ByteBuffer> {
+public final class BytesReadTrackingPublisher implements SdkHttpContentPublisher {
```

Comment:
Is there any reason we changed this class to implement SdkHttpContentPublisher?

Reply:
Yes, this is to enable the ByteReadTrackingPublisher to track the AsyncBody publisher of type SdkHttpContentPublisher : https://github.com/aws/aws-sdk-java-v2/blob/f6f2c251c285764a64a733cfb18e6bb2a2eae686/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MakeAsyncHttpRequestStage.java#L138. The same class, ByteReadTrackingPublisher is used to track requests and response